storage: don't truncate when probing active followers are present

tbg opened the pull request:
I need to clean up the first commit, putting the methods on `replica.mu` instead.

----

This is yet another root cause of Raft snapshots during import/restore. We
create a number of new ranges which may experience Raft leadership changes
early on and ingest SSTs at the same time. This leads to a situation in
which we have followers whose acknowledged log index isn't known, which in
turn leads to bad truncation decisions. Respect such followers (which are
usually only in this state for very brief amounts of time).

Release note: None

tbg committed:
storage: factor out lastUpdateMap's accessors

This is a more natural place to put them a) because there's no reason to
have them sit on Replica and b) because this will allow me to write
better code when I query this map in the log truncation code.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32439)
<!-- Reviewable:end -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-:-LRXXRUq2tdsPaDyHMy8:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/raft_log_queue.go, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRXX7kc6iNBIwhIr59f:-LRXX7kc6iNBIwhIr59g:b-cpzzir) ([raw file](https://github.com/cockroachdb/cockroach/blob/a542a0d70eea15f23e3ea5772af60f228b2532fe/pkg/storage/raft_log_queue.go#L160)):*
> ```Go
> 		// (including only resetting when the follower resumes replicating).
> 		pr.PendingSnapshot = pendingSnapshotIndex
> 	}
> ```

I think it is worthwhile to unit test this code as well. Perhaps it should be pulled out into a separate routine to ease the testing (i.e. something that takes a Raft status, descriptor and lastUpdateTimes map and returns the updated status).
___
*[pkg/storage/raft_log_queue.go, line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRXWN2W414CGAQvW979:-LRXWN2W414CGAQvW97A:bj9mi0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a542a0d70eea15f23e3ea5772af60f228b2532fe/pkg/storage/raft_log_queue.go#L179)):*
> ```Go
> 	truncatableIndexChosenViaQuorumIndex     = "quorum"
> 	truncatableIndexChosenViaFollowers       = "followers"
> 	truncatableIndexChosenViaProbingFollower = "probing-follower"
> ```

`s/probing-follower/probing follower/g` for consistency with the other two word reasons (e.g. `pending snapshot`, `first index`, etc).


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32439)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/raft_log_queue.go, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRXX7kc6iNBIwhIr59f:-LRg6KIR3R7BJiMUSAZ3:b-r4q3vl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a542a0d70eea15f23e3ea5772af60f228b2532fe/pkg/storage/raft_log_queue.go#L160)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think it is worthwhile to unit test this code as well. Perhaps it should be pulled out into a separate routine to ease the testing (i.e. something that takes a Raft status, descriptor and lastUpdateTimes map and returns the updated status).
</blockquote></details>

Thank you for making me write that unit test. I wasn't writing `pr` back into `prs` above, so in effect all that fancy code wasn't doing anything (`RecentActive` would've always been true).
___
*[pkg/storage/raft_log_queue.go, line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRXWN2W414CGAQvW979:-LRg2rKr5MiSiVj7wQQ1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a542a0d70eea15f23e3ea5772af60f228b2532fe/pkg/storage/raft_log_queue.go#L179)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/probing-follower/probing follower/g` for consistency with the other two word reasons (e.g. `pending snapshot`, `first index`, etc).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32439)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/raft_log_queue.go, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRXX7kc6iNBIwhIr59f:-LRh69YW9FEDgzZVg_gL:b-ms1byh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a542a0d70eea15f23e3ea5772af60f228b2532fe/pkg/storage/raft_log_queue.go#L160)):*
<details><summary><i>Previously, tbg (Tobias Schottdorf) wrote…</i></summary><blockquote>

Thank you for making me write that unit test. I wasn't writing `pr` back into `prs` above, so in effect all that fancy code wasn't doing anything (`RecentActive` would've always been true).
</blockquote></details>

Glad to be of service.
___
*[pkg/storage/raft_log_queue.go, line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRh0Re_0Vy9Knrd_TtS:-LRh0Re_0Vy9Knrd_TtT:b-mv246x) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6f3967ef5168dd51cab2b48c6eb26284c14852/pkg/storage/raft_log_queue.go#L146)):*
> ```Go
> 	// true since we don't use CheckQuorum) with our own activity check.
> 	r.mu.RLock()
> 	updateRaftProgressFromActivity(ctx, raftStatus.Progress, r.descRLocked().Replicas, r.mu.lastUpdateTimes, now)
> ```

Nit: wrap this line at 100 columns.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
PTAL. I had to make some more minor adjustments, including one that prompted an upstream PR (with a workaround for the time being).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-:-LRlAMS62NzDhiKNVO79:b-e4vcey)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/raft_log_queue.go, line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRh0Re_0Vy9Knrd_TtS:-LRh73eG4KaNuMutSfdo:bi6lrhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6f3967ef5168dd51cab2b48c6eb26284c14852/pkg/storage/raft_log_queue.go#L146)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: wrap this line at 100 columns.
</blockquote></details>

Done (tests are failing left and right now anyway), but generally, I'd prefer that we either put this into `crlfmt` or leave these alone in reviews. Do you know if we've tried and failed?


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32439)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/raft_log_queue.go, line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRh0Re_0Vy9Knrd_TtS:-LRlQDDv5wvWWXmKUGSw:b-hpkbtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6f3967ef5168dd51cab2b48c6eb26284c14852/pkg/storage/raft_log_queue.go#L146)):*
<details><summary><i>Previously, tbg (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done (tests are failing left and right now anyway), but generally, I'd prefer that we either put this into `crlfmt` or leave these alone in reviews. Do you know if we've tried and failed?
</blockquote></details>

I don't know if we've tried. The primary reason I care about the line length limit is that it makes reviewing in reviewable easier. 
___
*[pkg/storage/replica.go, line 456 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRlPdP10KTBR49zU0wW:-LRlPdP10KTBR49zU0wX:b-6rnjyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/297a14cc4847a47e83ded5d023b8db676bc3c307/pkg/storage/replica.go#L456)):*
> ```Go
> 		// TODO(tschottdorf): keeping a map on each replica seems to be
> 		// overdoing it. We should map the replicaID to a NodeID and then use
> 		// node liveness (or any sensible measure of the peer being around).
> ```

The motivation for tracking activity per-range is that a node could be active while a replica on that node is stuck for some reason. Not sure if this is sufficient motivation, but there it is. Would be good to expand this TODO with this additional detail so that we don't forget about it whenever the TODO is addressed.
___
*[pkg/storage/replica.go, line 5260 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRlOsiU3mV_1wg5Z0m5:-LRlOsiU3mV_1wg5Z0m6:b-k7fe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/297a14cc4847a47e83ded5d023b8db676bc3c307/pkg/storage/replica.go#L5260)):*
> ```Go
> 	fromReplica, fromErr := r.getReplicaDescriptorByIDRLocked(roachpb.ReplicaID(msg.From), r.mu.lastToReplica)
> 	toReplica, toErr := r.getReplicaDescriptorByIDRLocked(roachpb.ReplicaID(msg.To), r.mu.lastFromReplica)
> 	if isHB := msg.Type == raftpb.MsgHeartbeat; isHB {
> ```

The `isHB` variable is inhibiting readability. I'm guessing it is detritus left over from earlier work. 
___
*[pkg/storage/replica.go, line 5267 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRlPKyB6ADSlXcSkMQ4:-LRlPKyB6ADSlXcSkMQ5:beryi8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/297a14cc4847a47e83ded5d023b8db676bc3c307/pkg/storage/replica.go#L5267)):*
> ```Go
> 		// them into the local Raft group. The leader won't hit that path, so we
> 		// update it whenever it sends a heartbeat. In effect, this makes sure
> 		// it always sees itself as alive.
> ```

I wonder if it would be better to special-case the leader when using `lastUpdateTimes`. I don't have the usage paged-in, to feel free to ignore this idea if it is misguided.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR! I'm going to whack this with a few testraces on my worker to make up for the fact that TC won't really testrace many potentially affected tests in CI.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-:-LRlU2j-6kjn7DqFT7Zj:bd6s583)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica.go, line 456 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRlPdP10KTBR49zU0wW:-LRlT0xaC-zdnwRXXfGg:b-vavih2) ([raw file](https://github.com/cockroachdb/cockroach/blob/297a14cc4847a47e83ded5d023b8db676bc3c307/pkg/storage/replica.go#L456)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The motivation for tracking activity per-range is that a node could be active while a replica on that node is stuck for some reason. Not sure if this is sufficient motivation, but there it is. Would be good to expand this TODO with this additional detail so that we don't forget about it whenever the TODO is addressed.
</blockquote></details>

Right, but then we're already screwed because node liveness doesn't capture any of that neither. Updating the comment regardless.
___
*[pkg/storage/replica.go, line 5260 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRlOsiU3mV_1wg5Z0m5:-LRlTdhGByWxKWnfPlPr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/297a14cc4847a47e83ded5d023b8db676bc3c307/pkg/storage/replica.go#L5260)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `isHB` variable is inhibiting readability. I'm guessing it is detritus left over from earlier work. 
</blockquote></details>

Done.
___
*[pkg/storage/replica.go, line 5267 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32439#-LRlPKyB6ADSlXcSkMQ4:-LRlTeQ_8XWrQ8VcWpXs:bwwfz7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/297a14cc4847a47e83ded5d023b8db676bc3c307/pkg/storage/replica.go#L5267)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I wonder if it would be better to special-case the leader when using `lastUpdateTimes`. I don't have the usage paged-in, to feel free to ignore this idea if it is misguided.
</blockquote></details>

I considered that, but I think it's ultimately a bad idea to special case because every caller needs to do the right thing with it. For example, the quota pool is completely oblivious to these changes and would omit the leader in its computations. That is fine, but was probably not intentional. There are various other places in the code which compute ad-hoc quorum indexes. I want them to avoid this same kind of pitfall as they get made smarter in the future.


<!-- Sent from Reviewable.io -->

tbg committed:
storage: don't truncate when probing active followers are present

This is yet another root cause of Raft snapshots during import/restore.
We create a number of new ranges which may experience Raft leadership
changes early on and ingest SSTs at the same time. This leads to a
situation in which we have followers whose acknowledged log index isn't
known, which in turn leads to bad truncation decisions. Respect such
followers (which are usually only in this state for very brief amounts
of time).

Release note: None

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32439)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

tbg commented:
Wow, turns out running the logictests under `testrace` is a suicidal idea.

bors r=petermattis
